src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -162,17 +164,108 @@ MdnsContexts MdnsContexts::sInstance;
+        sdCtx->Finalize(err);
+    }
+};
+
 static void OnRegister(DNSServiceRef sdRef, DNSServiceFlags flags, DNSServiceErrorType err, const char * name, const char * type,
                        const char * domain, void * context)
 {
-    ChipLogDetail(Discovery, "Mdns: %s name: %s, type: %s, domain: %s, flags: %d", __func__, name, type, domain, flags);
-
+    ChipLogDetail(Discovery, "Mdns: %s name: %s, type: %s, domain: %s, flags: %d Error %d", __func__, name, type, domain, flags,
+                  err);
     auto sdCtx = reinterpret_cast<RegisterContext *>(context);
     sdCtx->Finalize(err);
 };
 
+CHIP_ERROR RegisterService(DNSServiceRef sdRef, DNSServiceFlags flags, uint32_t interfaceId, const char * name, const char * type,
+                           const char * domain, const char * hostname, uint16_t port, uint16_t txtLen, const void * txtRecord,
+                           DNSServiceRegisterReply callback, RegisterContext * sdCtx)
+{
+    auto err = DNSServiceRegister(&sdRef, flags, interfaceId, name, type, domain, hostname, ntohs(port), txtLen, txtRecord,
+                                  callback, sdCtx);
+    ChipLogProgress(Discovery, "Registering service %s on port %u with type: %s on interface id: %u hostname %s error: %d" PRIu32,
+                    name, port, type, interfaceId, hostname, err);
+    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
+    return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+}
+
+CHIP_ERROR RegisterRecord(RegisterRecordContext * registerRecordCtx, const char * type, uint32_t interfaceId, const char * hostname)
+{
+    DNSServiceRef sdRef;
+    auto err = DNSServiceCreateConnection(&sdRef);
+    VerifyOrReturnError((kDNSServiceErr_NoError == err || sdRef != nullptr), registerRecordCtx->Finalize(err));
+
+    auto error = MdnsContexts::GetInstance().Add(registerRecordCtx, sdRef);
+    VerifyOrReturnError(CHIP_NO_ERROR == error, error);
+    // Get all the interface addresses
+    struct ifaddrs * ifaddr;
+    auto ret = getifaddrs(&ifaddr);
+    if (ret)
+    {
+        ChipLogError(Discovery, "Getting interface addresses failed %d.", ret);
+        freeifaddrs(ifaddr);
+    }
+    VerifyOrReturnError(ret == 0, registerRecordCtx->Finalize(ret));
+
+    // Call DNSServiceRegisterRecord for each interface if interface id is of type any so we can register the address for each
+    // interface Otherwise just register record for the given interface id
+    bool isInterfaceIndexAny = (interfaceId == kDNSServiceInterfaceIndexAny);
+    for (struct ifaddrs * ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next)
+    {
+        if (ifa->ifa_addr->sa_family == AF_INET)
+        {
+            struct sockaddr_in * inaddr = (struct sockaddr_in *) ifa->ifa_addr;
+            bool shouldRegisterRecord =
+                (isInterfaceIndexAny || (!isInterfaceIndexAny && (if_nametoindex(ifa->ifa_name) == interfaceId)));
```

Comment:
```
bool shouldRegisterRecord =
                 (isInterfaceIndexAny || (!isInterfaceIndexAny && (if_nametoindex(ifa->ifa_name) == interfaceId)));
if (!shouldRegisterRecord)
{
    continue;
}
```
This can probably be done "before" looking at the `sa_family` and it would avoid to duplicate this check.

Reply:
done

Reply:
done

Reply:
:) yes